Add tests for PELT

mdigiorgio opened the pull request:
This PR adds a couple of tests for making sure PELT signals (right, util_avg_task only is tested) works as expected.

The tests rely on the PELT simulator that @derkling is adding to bart (see https://github.com/mdigiorgio/bart/tree/pelt_simulator). As a consequence, we should have that merged before we accept this.

I'm posting this now to get some initial feedback on the approach, that's why I set it as WIP.

bjackman commented on tests/eas/load_tracking.py:
```diff
@@ -59,16 +59,23 @@ def setUpClass(cls, *args, **kwargs):
 
     @classmethod
-    def get_wload(cls, cpu):
+    def get_wload(cls, cpu, duty_cycle):
```
Call it `duty_cycle_pct` for consistency?

bjackman commented on tests/eas/load_tracking.py:
```diff
@@ -172,6 +175,81 @@ def get_signal_mean(self, experiment, signal,
+        # Get instant of time when the task starts running
+        sw_df = trace.ftrace.sched_switch.data_frame
+        pid = sw_df[sw_df.prev_comm == task].prev_pid.iloc[0]
+        return sw_df[sw_df.next_pid == pid].index[0]
```
`LisaTest` has a `get_start_time` method - can you integrate this with the logic used there, and put it next to it?

bjackman commented on tests/eas/load_tracking.py:
```diff
@@ -172,6 +175,81 @@ def get_signal_mean(self, experiment, signal,
+        """
+        Get simulated PELT signal and the periodic task used to model it.
+
+        :returns:
```
Should probably specify this is a __tuple__ of the 3 things

bjackman commented on tests/eas/load_tracking.py:
```diff
@@ -165,6 +175,81 @@ def get_signal_mean(self, experiment, signal,
+        updates_since_start = int(time_since_start / (peltsim._sample_us / 1e6))
+        util = signal_df.iloc[0]
+        for i in range(updates_since_start):
+            util = (util - peltsim._geom_u) / peltsim._geom_y
```
This looks like something that belongs in the `Simulator`.

bjackman commented on tests/eas/load_tracking.py:
```diff
@@ -323,3 +408,163 @@ def test_task_util_avg(self, experiment, tasks):
+        """
+        Test PELT util_avg stable range for a 50% periodic task
+        """
+        return self._test_range(experiment, tasks, 'util_avg')
```
Are the `_test_range` and `_test_behavoiur` helpers actually needed? Looks like you can just put the code directly in the test method.

mdigiorgio replied:
They can be used for testing `load_avg` as well, that's why I kept those separate.

mdigiorgio commented:
Addressed @bjackman's comments. I'll also push an update on the commits for the PELT simulator to move the code that estimates the initial value.

bjackman commented:
Not sure what the travis failure is about, have restarted it.

bjackman commented:
For the record, the Travis failure is because @mdigiorgio accidentally pushed a pelt-test branch to the ARM-Software repo - at this point Travis triggered a job to test that branch - but then deleted the branch. So Travis is trying to check out a non-existent branch. We can just ignore it - the "continuous-integration/travis-ci/pr" check is the one we care about.

mdigiorgio commented on tests/eas/load_tracking.py:
```diff
@@ -42,7 +45,7 @@ class _LoadTrackingBase(LisaTest):
         },
         # cgroups required by freeze_userspace flag
-        'modules': ['cpufreq', 'cgroups'],
+        'modules': ['cpufreq', 'cgroups', 'bl'],
```
That's what I hope, but I'll give it a spin on hikey620 to be sure

mdigiorgio replied:
Changed it to get a big CPU from rt-app calibration values as it fails when the platform does not have the bl module.

mdigiorgio commented:
Addressed comments and added tests for `load_avg_task`.

mdigiorgio commented:
Added some tests for taskgroups.

mdigiorgio commented:
Just pushed an updated version of the patchset. News are the following:
- PELT simulator has been merged in BART
- Added a data structure for creating hierarchies of tasks and taskgroups
- Added tests for utilization of taskgroups using the new `SchedEntity` class
